File: Intercom/app/src/main/java/com/twiceyuan/intercom/common/Encrypt.java
package com.twiceyuan.intercom.common;

import java.io.UnsupportedEncodingException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Created by twiceYuan on 8/9/16.
 * Email: i@twiceyuan.com
 * Site: http://twiceyuan.com
 */
public class Encrypt {

    public static String md5(String origin) {
        try {
            return byteArrayToHex(MessageDigest.getInstance("MD5").digest(origin.getBytes("UTF-8")));
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
            return null;
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
            return null;
        }
    }

    public static String byteArrayToHex(byte[] byteArray) {
        char[] hexDigits = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
        char[] resultCharArray = new char[byteArray.length * 2];
        int index = 0;
        for (byte b : byteArray) {
            resultCharArray[index++] = hexDigits[b >>> 4 & 0xf];
            resultCharArray[index++] = hexDigits[b & 0xf];
        }
        return new String(resultCharArray);
    }
}

File: Intercom/app/src/main/java/com/twiceyuan/intercom/common/Encrypt.kt
package com.twiceyuan.intercom.common

import java.io.UnsupportedEncodingException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * Created by twiceYuan on 8/9/16.
 * Email: i@twiceyuan.com
 * Site: http://twiceyuan.com
 */
object Encrypt {

    fun md5(origin: String): String? {
        try {
            return byteArrayToHex(MessageDigest.getInstance("MD5").digest(origin.toByteArray(charset("UTF-8"))))
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            return null
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
            return null
        }

    }

    fun byteArrayToHex(byteArray: ByteArray): String {
        val hexDigits = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')
        val resultCharArray = CharArray(byteArray.size * 2)
        var index = 0
        for (b in byteArray) {
            // Kotlin 中位运算仅限于 Int 和 Long 因此需要先显式转换为 Int 型
            resultCharArray[index++] = hexDigits[b.toInt().ushr(4) and 0xf]
            resultCharArray[index++] = hexDigits[b.toInt() and 0xf]
        }
        return String(resultCharArray)
    }
}
